File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/poker/poker.go
package poker

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

type Card struct {
	Rank int
	Suit rune
}

type HandRank struct {
	Type   int
	Values []int
}

const (
	HighCard = iota
	OnePair
	TwoPairs
	ThreeOfAKind
	Straight
	Flush
	FullHouse
	FourOfAKind
	StraightFlush
)

func parseHand(hand string) ([]Card, error) {
	cards := strings.Split(hand, " ")
	if len(cards) != 5 {
		return nil, fmt.Errorf("invalid hand: %s", hand)
	}

	var result []Card
	for _, cardStr := range cards {
		card, err := parseCard(cardStr)
		if err != nil {
			return nil, err
		}
		result = append(result, card)
	}
	return result, nil
}

func parseCard(cardStr string) (Card, error) {
	runes := []rune(cardStr)
	
	// Handle 10 case with proper rune handling
	if len(runes) == 3 && runes[0] == '1' && runes[1] == '0' {
		return Card{Rank: 10, Suit: runes[2]}, nil
	}
	
	if len(runes) != 2 {
		return Card{}, fmt.Errorf("invalid card: %s", cardStr)
	}

	rankChar := runes[0]
	suit := runes[1]

	var rank int
	switch rankChar {
	case 'J':
		rank = 11
	case 'Q':
		rank = 12
	case 'K':
		rank = 13
	case 'A':
		rank = 14
	default:
		var err error
		rank, err = strconv.Atoi(string(rankChar))
		if err != nil || rank < 2 || rank > 9 {
			return Card{}, fmt.Errorf("invalid rank: %c", rankChar)
		}
	}

	return Card{Rank: rank, Suit: suit}, nil
}

func evaluateHand(hand []Card) HandRank {
	sorted := make([]Card, len(hand))
	copy(sorted, hand)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Rank < sorted[j].Rank })

	isFlush := true
	suit := sorted[0].Suit
	for _, c := range sorted[1:] {
		if c.Suit != suit {
			isFlush = false
			break
		}
	}

	isStraight := true
	for i := 1; i < len(sorted); i++ {
		if sorted[i].Rank != sorted[i-1].Rank+1 {
			if i == len(sorted)-1 && sorted[0].Rank == 2 && sorted[1].Rank == 3 &&
				sorted[2].Rank == 4 && sorted[3].Rank == 5 && sorted[4].Rank == 14 {
				// Ace-low straight (5-high)
				return HandRank{
					Type:   Straight,
					Values: []int{5},
				}
			}
			isStraight = false
			break
		}
	}

	// Ace-low straight check
	if !isStraight && sorted[0].Rank == 2 && sorted[1].Rank == 3 && sorted[2].Rank == 4 &&
		sorted[3].Rank == 5 && sorted[4].Rank == 14 {
		isStraight = true
		sorted = []Card{sorted[0], sorted[1], sorted[2], sorted[3], {Rank: 1, Suit: sorted[4].Suit}}
		sort.Slice(sorted, func(i, j int) bool { return sorted[i].Rank < sorted[j].Rank })
	}

	rankCount := make(map[int]int)
	for _, c := range sorted {
		rankCount[c.Rank]++
	}

	var pairs, trips, quad []int
	var values []int
	for rank, count := range rankCount {
		switch count {
		case 2:
			pairs = append(pairs, rank)
		case 3:
			trips = append(trips, rank)
		case 4:
			quad = append(quad, rank)
		}
	}
	sort.Sort(sort.Reverse(sort.IntSlice(pairs)))
	sort.Sort(sort.Reverse(sort.IntSlice(trips)))

	// Determine hand type
	switch {
	case isStraight && isFlush:
		return HandRank{
			Type:   StraightFlush,
			Values: []int{sorted[len(sorted)-1].Rank},
		}

	case len(quad) == 1:
		kicker := 0
		for _, c := range sorted {
			if c.Rank != quad[0] {
				kicker = c.Rank
			}
		}
		return HandRank{
			Type:   FourOfAKind,
			Values: []int{quad[0], kicker},
		}

	case len(trips) == 1 && len(pairs) == 1:
		return HandRank{
			Type:   FullHouse,
			Values: []int{trips[0], pairs[0]},
		}

	case isFlush:
		values = make([]int, 5)
		for i := range sorted {
			values[i] = sorted[len(sorted)-1-i].Rank
		}
		return HandRank{
			Type:   Flush,
			Values: values,
		}

	case isStraight:
		return HandRank{
			Type:   Straight,
			Values: []int{sorted[len(sorted)-1].Rank},
		}

	case len(trips) == 1:
		kickers := make([]int, 0)
		for _, c := range sorted {
			if c.Rank != trips[0] {
				kickers = append(kickers, c.Rank)
			}
		}
		sort.Sort(sort.Reverse(sort.IntSlice(kickers)))
		return HandRank{
			Type:   ThreeOfAKind,
			Values: append([]int{trips[0]}, kickers...),
		}

	case len(pairs) >= 2:
		kicker := 0
		for _, c := range sorted {
			if c.Rank != pairs[0] && c.Rank != pairs[1] {
				kicker = c.Rank
				break
			}
		}
		return HandRank{
			Type:   TwoPairs,
			Values: []int{pairs[0], pairs[1], kicker},
		}

	case len(pairs) == 1:
		kickers := make([]int, 0)
		for _, c := range sorted {
			if c.Rank != pairs[0] {
				kickers = append(kickers, c.Rank)
			}
		}
		sort.Sort(sort.Reverse(sort.IntSlice(kickers)))
		return HandRank{
			Type:   OnePair,
			Values: append([]int{pairs[0]}, kickers...),
		}

	default:
		values = make([]int, 5)
		for i := range sorted {
			values[i] = sorted[len(sorted)-1-i].Rank
		}
		return HandRank{
			Type:   HighCard,
			Values: values,
		}
	}
}

func compareHandRanks(a, b HandRank) int {
	if a.Type != b.Type {
		return a.Type - b.Type
	}
	for i := range a.Values {
		if a.Values[i] != b.Values[i] {
			return a.Values[i] - b.Values[i]
		}
	}
	return 0
}

func BestHand(hands []string) ([]string, error) {
	var maxRank HandRank
	var bestHands []string

	for _, handStr := range hands {
		hand, err := parseHand(handStr)
		if err != nil {
			return nil, err
		}

		currentRank := evaluateHand(hand)

		if len(bestHands) == 0 {
			maxRank = currentRank
			bestHands = []string{handStr}
			continue
		}

		comparison := compareHandRanks(currentRank, maxRank)
		if comparison > 0 {
			maxRank = currentRank
			bestHands = []string{handStr}
		} else if comparison == 0 {
			bestHands = append(bestHands, handStr)
		}
	}

	return bestHands, nil
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/java/exercises/practice/poker/src/main/java/Poker.java
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.Arrays;
import java.util.stream.Collectors;

class Poker {
    private final List<String> hands;
    
    private enum HandType {
        STRAIGHT_FLUSH, FOUR_OF_A_KIND, FULL_HOUSE, FLUSH, STRAIGHT, 
        THREE_OF_A_KIND, TWO_PAIR, ONE_PAIR, HIGH_CARD
    }

    private static class HandRank {
        final HandType type;
        final List<Integer> key;
        
        HandRank(HandType type, List<Integer> key) {
            this.type = type;
            this.key = key;
        }
    }

    private static class Card {
        final int rankValue;
        final char suit;

        Card(int rankValue, char suit) {
            this.rankValue = rankValue;
            this.suit = suit;
        }
    }

    Poker(List<String> hand) {
        this.hands = new ArrayList<>(hand);
    }

    List<String> getBestHands() {
        List<HandRank> handRanks = new ArrayList<>();
        for (String hand : hands) {
            List<Card> cards = parseHand(hand);
            handRanks.add(evaluateHand(cards));
        }

        int maxRank = handRanks.stream()
            .mapToInt(r -> r.type.ordinal())
            .min()
            .orElse(-1);

        List<HandRank> maxHandRanks = handRanks.stream()
            .filter(r -> r.type.ordinal() == maxRank)
            .collect(Collectors.toList());

        List<HandRank> bestRanks = compareSameRank(maxHandRanks);

        List<String> bestHands = new ArrayList<>();
        for (int i = 0; i < handRanks.size(); i++) {
            if (bestRanks.contains(handRanks.get(i))) {
                bestHands.add(hands.get(i));
            }
        }
        return bestHands;
    }

    private List<Card> parseHand(String hand) {
        List<Card> cards = new ArrayList<>();
        for (String cardStr : hand.split(" ")) {
            String rank = cardStr.substring(0, cardStr.length() - 1);
            char suit = cardStr.charAt(cardStr.length() - 1);
            int value;
            switch (rank) {
                case "J": value = 11; break;
                case "Q": value = 12; break;
                case "K": value = 13; break;
                case "A": value = 14; break;
                default: value = Integer.parseInt(rank);
            }
            cards.add(new Card(value, suit));
        }
        Collections.sort(cards, (a, b) -> b.rankValue - a.rankValue);
        return cards;
    }

    private HandRank evaluateHand(List<Card> cards) {
        boolean isFlush = cards.stream().map(c -> c.suit).distinct().count() == 1;
        boolean isStraight = true;
        
        // Check normal straight
        for (int i = 0; i < cards.size() - 1; i++) {
            if (cards.get(i).rankValue - cards.get(i + 1).rankValue != 1) {
                isStraight = false;
                break;
            }
        }
        
        // Check low Ace straight (A-2-3-4-5)
        if (!isStraight && cards.get(0).rankValue == 14 
                && cards.get(1).rankValue == 5 
                && cards.get(2).rankValue == 4 
                && cards.get(3).rankValue == 3 
                && cards.get(4).rankValue == 2) {
            isStraight = true;
            Collections.rotate(cards, 1); // Rotate to make 5 the first card
        }

        Map<Integer, Integer> rankCount = new HashMap<>();
        for (Card c : cards) {
            rankCount.put(c.rankValue, rankCount.getOrDefault(c.rankValue, 0) + 1);
        }

        List<Integer> pairs = new ArrayList<>();
        int threeOfAKind = 0;
        int fourOfAKind = 0;
        for (Map.Entry<Integer, Integer> entry : rankCount.entrySet()) {
            if (entry.getValue() == 4) {
                fourOfAKind = entry.getKey();
            } else if (entry.getValue() == 3) {
                threeOfAKind = entry.getKey();
            } else if (entry.getValue() == 2) {
                pairs.add(entry.getKey());
            }
        }

        if (isStraight && isFlush) {
            return new HandRank(HandType.STRAIGHT_FLUSH, getHighCards(cards, 5));
        } else if (fourOfAKind > 0) {
            List<Integer> key = new ArrayList<>();
            key.add(fourOfAKind);
            key.add(cards.stream().filter(c -> c.rankValue != fourOfAKind).findFirst().get().rankValue);
            return new HandRank(HandType.FOUR_OF_A_KIND, key);
        } else if (threeOfAKind > 0 && !pairs.isEmpty()) {
            List<Integer> key = new ArrayList<>();
            key.add(threeOfAKind);
            key.add(pairs.get(0));
            return new HandRank(HandType.FULL_HOUSE, key);
        } else if (isFlush) {
            return new HandRank(HandType.FLUSH, getHighCards(cards, 5));
        } else if (isStraight) {
            return new HandRank(HandType.STRAIGHT, getHighCards(cards, 5));
        } else if (threeOfAKind > 0) {
            List<Integer> key = new ArrayList<>();
            key.add(threeOfAKind);
            key.addAll(getHighCards(cards, 2, threeOfAKind));
            return new HandRank(HandType.THREE_OF_A_KIND, key);
        } else if (pairs.size() == 2) {
            List<Integer> key = new ArrayList<>();
            key.add(Math.max(pairs.get(0), pairs.get(1)));
            key.add(Math.min(pairs.get(0), pairs.get(1)));
            key.addAll(getHighCards(cards, 1, pairs.get(0), pairs.get(1)));
            return new HandRank(HandType.TWO_PAIR, key);
        } else if (pairs.size() == 1) {
            List<Integer> key = new ArrayList<>();
            key.add(pairs.get(0));
            key.addAll(getHighCards(cards, 3, pairs.get(0)));
            return new HandRank(HandType.ONE_PAIR, key);
        } else {
            return new HandRank(HandType.HIGH_CARD, getHighCards(cards, 5));
        }
    }

    private List<Integer> getHighCards(List<Card> cards, int count) {
        return cards.stream()
            .map(c -> c.rankValue)
            .limit(count)
            .collect(Collectors.toList());
    }

    private List<Integer> getHighCards(List<Card> cards, int count, Integer... exclude) {
        List<Integer> excluded = Arrays.asList(exclude);
        return cards.stream()
            .map(c -> c.rankValue)
            .filter(r -> !excluded.contains(r))
            .sorted(Collections.reverseOrder())
            .limit(count)
            .collect(Collectors.toList());
    }

    private List<HandRank> compareSameRank(List<HandRank> ranks) {
        List<HandRank> best = new ArrayList<>();
        for (HandRank rank : ranks) {
            if (best.isEmpty()) {
                best.add(rank);
                continue;
            }
            int cmp = compareKeys(rank.key, best.get(0).key);
            if (cmp > 0) {
                best.clear();
                best.add(rank);
            } else if (cmp == 0) {
                best.add(rank);
            }
        }
        return best;
    }

    private int compareKeys(List<Integer> key1, List<Integer> key2) {
        for (int i = 0; i < key1.size(); i++) {
            int cmp = Integer.compare(key1.get(i), key2.get(i));
            if (cmp != 0) {
                return cmp;
            }
        }
        return 0;
    }
}
